File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples = (sum .) . multiples

multiples :: [Integer] -> Integer -> [Integer]
multiples fact limit = [x| x <- [1..limit-1], any (multipleOf x) fact']
  where
    fact' = filter (/=0) fact
    multipleOf = ((0 ==) .) . mod

File: cpp/sum-of-multiples/sum_of_multiples.cpp
#include "sum_of_multiples.h"

#include <algorithm>
#include <vector>

int sum_of_multiples::to(const std::vector<int> &factors, int limit) {
    int sum = 0;
    for (int i = 0; i < limit; i++) {
        if (std::any_of(factors.cbegin(), factors.cend(),
                        [&i](const int f) { return f > 0 && i % f == 0; })) {
            sum += i;
        }
    }

    return sum;
}
